misk/src/main/kotlin/misk/client/PeerClientFactory.kt

```diff
@@ -0,0 +1,56 @@
+package misk.client
+
+import misk.clustering.Cluster
+import misk.config.AppName
+import misk.security.cert.X500Name
+import misk.web.jetty.JettyService
+import okhttp3.OkHttpClient
+import java.security.cert.X509Certificate
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Factory that creates [OkHttpClient]s for connecting to another instance of the same application
+ * running in the same cluster.
+ */
+@Singleton
+class PeerClientFactory @Inject constructor(
+  @AppName private val appName: String,
+  private val jetty: JettyService,
+  private val httpClientsConfig: HttpClientsConfig,
+  private val httpClientFactory: HttpClientFactory
+) {
+
+  init {
+    require(jetty.httpsServerUrl != null) { "SSL must be configured to use peer clients" }
+
+    // There must be web client config. The URL and Envoy config are ultimately ignored.
+    httpClientsConfig[appName] // This throws if config is missing
+  }
+
+  /** Get the base URL for calling the given peer cluster member. */
+  fun baseUrl(peer: Cluster.Member): String {
+    return "https://${peer.ipAddress}:${jetty.httpsServerUrl!!.port()}"
+  }
+
+  /**
+   * Get a client to call the given peer cluster member.
+   * This client will fail when calling different services, as determined by the OU in the certificate
+   * returned by the called service.
+   */
+  fun client(peer: Cluster.Member): OkHttpClient {
```

Comment:
what level does connection pooling happen at? Should the clients per peer be singletons too?

Reply:
Good question. `OkHttpClient` pools connections so it would make sense to for `PeerClientFactory` to cache them.

@swankjesse, I noticed the existing `HttpClientModule` doesn't scope clients as singletons. Any particular reason why to keep in mind?